File: backend/app/schemas/analytics.py
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class AnalyticsCourseOption(BaseModel):
    id: int
    name: str


class AnalyticsSummary(BaseModel):
    courses: int
    lessons: int
    attendance_total: int
    attendance_attended: int
    attendance_rate: float = Field(ge=0, le=1)
    feedback_count: int
    feedback_avg: Optional[float] = None


class AnalyticsTimeseriesPoint(BaseModel):
    date: str
    total: int
    attended: int
    attendance_rate: float = Field(ge=0, le=1)


class AnalyticsRatingBucket(BaseModel):
    rating: int = Field(ge=1, le=5)
    count: int = Field(ge=0)


class AnalyticsTopAbsentStudent(BaseModel):
    student_id: int
    student_name: str
    absent: int = Field(ge=0)
    total: int = Field(ge=0)
    absent_rate: float = Field(ge=0, le=1)


class AnalyticsCourseSummaryRow(BaseModel):
    course_id: int
    course_name: str
    lessons: int = Field(ge=0)
    attendance_rate: float = Field(ge=0, le=1)
    feedback_avg: Optional[float] = None
    feedback_count: int = Field(ge=0)


class AnalyticsOverview(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    role: str
    scope: str
    courses: List[AnalyticsCourseOption]
    summary: AnalyticsSummary
    timeseries: List[AnalyticsTimeseriesPoint]
    rating_distribution: List[AnalyticsRatingBucket]
    top_absent_students: List[AnalyticsTopAbsentStudent]
    course_summary: List[AnalyticsCourseSummaryRow]


class AnalyticsRiskRow(BaseModel):
    student_id: int
    student_name: str
    course_id: int
    course_name: str
    total_records: int = Field(ge=0)
    window_size: int = Field(ge=0)
    recent_absent_rate: float = Field(ge=0, le=1)
    absent_streak: int = Field(ge=0)
    risk_absent_next: float = Field(ge=0, le=1)
    model: str
    confidence: float = Field(ge=0, le=1)


class AnalyticsRiskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    role: str
    scope: str
    algorithm: str
    trained: bool
    training_samples: int = Field(ge=0)
    features: List[str]
    rows: List[AnalyticsRiskRow]

File: frontend/src/api/analytics.ts
import api from "./client";

export type AnalyticsScope = "auto" | "overall" | "personal";

export interface AnalyticsParams {
  course_id?: number;
  from_date?: string;
  to_date?: string;
  scope?: AnalyticsScope;
}

export interface AnalyticsCourseOption {
  id: number;
  name: string;
}

export interface AnalyticsSummary {
  courses: number;
  lessons: number;
  attendance_total: number;
  attendance_attended: number;
  attendance_rate: number;
  feedback_count: number;
  feedback_avg: number | null;
}

export interface AnalyticsTimeseriesPoint {
  date: string;
  total: number;
  attended: number;
  attendance_rate: number;
}

export interface AnalyticsRatingBucket {
  rating: number;
  count: number;
}

export interface AnalyticsTopAbsentStudent {
  student_id: number;
  student_name: string;
  absent: number;
  total: number;
  absent_rate: number;
}

export interface AnalyticsCourseSummaryRow {
  course_id: number;
  course_name: string;
  lessons: number;
  attendance_rate: number;
  feedback_avg: number | null;
  feedback_count: number;
}

export interface AnalyticsOverview {
  role: "admin" | "teacher" | "student";
  scope: "overall" | "personal";
  courses: AnalyticsCourseOption[];
  summary: AnalyticsSummary;
  timeseries: AnalyticsTimeseriesPoint[];
  rating_distribution: AnalyticsRatingBucket[];
  top_absent_students: AnalyticsTopAbsentStudent[];
  course_summary: AnalyticsCourseSummaryRow[];
}

export async function getAnalyticsOverview(params: AnalyticsParams = {}): Promise<AnalyticsOverview> {
  const response = await api.get("/analytics/overview", { params });
  return response.data as AnalyticsOverview;
}

export interface AnalyticsRiskParams extends AnalyticsParams {
  k?: number;
  limit?: number;
}

export interface AnalyticsRiskRow {
  student_id: number;
  student_name: string;
  course_id: number;
  course_name: string;
  total_records: number;
  window_size: number;
  recent_absent_rate: number;
  absent_streak: number;
  risk_absent_next: number;
  model: string;
  confidence: number;
}

export interface AnalyticsRiskResponse {
  role: "admin" | "teacher" | "student";
  scope: "overall" | "personal";
  algorithm: string;
  trained: boolean;
  training_samples: number;
  features: string[];
  rows: AnalyticsRiskRow[];
}

export async function getAnalyticsRisk(params: AnalyticsRiskParams = {}): Promise<AnalyticsRiskResponse> {
  const response = await api.get("/analytics/risk", { params });
  return response.data as AnalyticsRiskResponse;
}
